test/integration/upgrade/test_upgrade.py

```diff
@@ -0,0 +1,106 @@
+import re
+
+import pytest
+from vshard_cluster import VshardCluster
+
+from utils import get_tarantool_version, run_command_and_get_output
+
+tarantool_major_version, _ = get_tarantool_version()
+
+
+@pytest.mark.skipif(tarantool_major_version < 3,
+                    reason="skip test with cluster config for Tarantool < 3")
+def test_upgrade_all_replicasets(tt_cmd, tmp_path):
```

Comment:
All the test cases perform a no-op upgrade: the instances are already on the last database schema version. As result an implementation that doesn't call `box.schema.upgrade()` at all would pass this test. No-op `WaitLSN` would also pass it.

@psergee Should we add scenarios that perform a real upgrade from, say, a 2.11.1 schema? Or it is considered as too complicated in the given testing infrastructure?

I see two variants how to implement it:

* Save the pre-generated 2.11 snapshot into the repository, copy them into appropriate instance directories before starting the cluster.
* Downgrade the database schema to 2.11.1 on running instances before calling the `tt upgrade` command.

Dowgrading may need extra tricks before https://github.com/tarantool/tarantool/issues/10150 is solved if a cluster configuration is in use.

Reply:
Still in progress.